app/code/Magento/Store/Model/StoreManager.php

```diff
@@ -152,25 +152,37 @@ public function isSingleStoreMode()
     public function getStore($storeId = null)
     {
         if (!isset($storeId) || '' === $storeId || $storeId === true) {
-            if (null === $this->currentStoreId) {
-                \Magento\Framework\Profiler::start('store.resolve');
-                $this->currentStoreId = $this->storeResolver->getCurrentStoreId();
-                \Magento\Framework\Profiler::stop('store.resolve');
-            }
-            $storeId = $this->currentStoreId;
+            $storeId = $this->getCurrentStoreId();
         }
         if ($storeId instanceof \Magento\Store\Api\Data\StoreInterface) {
             return $storeId;
         }
 
-        $store = is_numeric($storeId)
-            ? $this->storeRepository->getById($storeId)
-            : $this->storeRepository->get($storeId);
+        if (is_numeric($storeId)) {
+            $store = $this->storeRepository->getById($storeId);
+        } elseif (key_exists($storeId, $this->storeRepository->getList())) {
+            $store = $this->storeRepository->get($storeId);
+        } else {
+            $store = $this->storeRepository->getById($this->getCurrentStoreId());
```

Comment:
I am a bit curious about this else. From what I can see `$storeId` will be the result of `$this->getCurrentStoreId()`. Is there any reason you have not just used `$storeId` here rather than calling `$this->getCurrentStoreId()`?

Reply:
Also from what I can see this thing could be done with:

```
if (is_numeric($storeId) || !key_exists($storeId, $this->storeRepository->getList())) {
    $store = $this->storeRepository->getById($storeId);
} else {
    $store = $this->storeRepository->get($storeId);
}
```

Or have I missed something out here?

Reply:
@dmanners `$storeId` may be passed from outside as a method parameter. And this is actually the issue, as there are several places where `___store` query parameter just fetched from a query object and passed to store manager. Store resolver already implements a logic for correct handling of the invalid store code or id.

I'm worrying about this implementation as at this point we do not know exactly that `$storeId` value comes from `___store` query parameter so any code issue will be silently handled without any notification and may lead to unpredictable result and hard to detect bugs.

I think we should find a solution in which only value of `___store` query parameter will be affected.